plugins/extractors/postgres/postgres.go

```diff
@@ -0,0 +1,201 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+
+	_ "github.com/lib/pq"
+	"github.com/odpf/meteor/core/extractor"
+	"github.com/odpf/meteor/plugins"
+	"github.com/odpf/meteor/proto/odpf/meta"
+	"github.com/odpf/meteor/proto/odpf/meta/facets"
+	"github.com/odpf/meteor/utils"
+)
+
+type Config struct {
+	UserID   string `mapstructure:"user_id" validate:"required"`
+	Password string `mapstructure:"password" validate:"required"`
+	Host     string `mapstructure:"host" validate:"required"`
+	Database string `mapstructure:"database" default:"postgres"`
+	Exclude  string `mapstructure:"exclude" default:"information_schema,root,postgres"`
+}
+
+type Extractor struct {
+	logger plugins.Logger
+}
+
+// Extract collects metdata from the source. Metadata is collected through the out channel
+func (e *Extractor) Extract(ctx context.Context, config map[string]interface{}, out chan<- interface{}) (err error) {
+
+	// Build and validate config received from receipe
+	var cfg Config
+	if err := utils.BuildConfig(config, &cfg); err != nil {
+		return extractor.InvalidConfigError{}
+	}
+
+	// Create database connection
+	db, err := connection(cfg, cfg.Database)
+	if err != nil {
+		return err
+	}
+
+	// Get list of databases
+	dbs, err := e.getDatabases(cfg, db)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	// Iterate through all tables and databases
+	for _, database := range dbs {
+		// Open a new connection to the given database to collect
+		// tables information without this default database
+		// information will be returned
+		db, err := connection(cfg, database)
+		if err != nil {
+			e.logger.Error(err)
+			continue
+		}
+		tables, err := e.getTables(db, database)
+		if err != nil {
+			e.logger.Error(err)
+			continue
+		}
+
+		for _, table := range tables {
+			result, err := e.getTableMetadata(db, database, table)
+			if err != nil {
+				e.logger.Error(err)
+				continue
+			}
+			// Publish metadata to channel
+			out <- result
+		}
+	}
+
+	return nil
+}
+
+func (e *Extractor) getDatabases(cfg Config, db *sql.DB) (list []string, err error) {
+	res, err := db.Query("SELECT datname FROM pg_database WHERE datistemplate = false;")
+	if err != nil {
+		return nil, err
+	}
+
+	for res.Next() {
+		var database string
+		res.Scan(&database)
+		if exclude(cfg, database) {
```

Comment:
this exclude will split a string and transform it into array for every table in every database.
should transform it once and store the array to the extractor struct?

also I was thinking we could probably use map instead of array for quicker check since Exclude is now configurable. (edge cases where users excludes lots of table)

Reply:
Yeah, so we can split the Exclude string once and map the values? That will be a one time operation, so it should be efficient?

Reply:
> Also any reason for renaming the `extractor.go` to `postgres.go`?? I think our package name is `postgres` already so I thought `extractor.go` was fine seeing the intent of that code.

File searching in the editor becomes very tricky with all files named extractors. And general Go practice is also to use the main file of the package as the same name.

Reply:
@GrayFlash yes, transforming the exclude string to a map (e.g. map[string]bool) will be a once for one extraction.
Also when checking each of table if it is excluded will be constant compared to if it is an array (we have to loop through all the items in the array for almost every table).